util/file_reader_writer.cc

```diff
@@ -743,13 +735,147 @@ class ReadaheadRandomAccessFile : public RandomAccessFile {
     return s;
   }
 
-  std::unique_ptr<RandomAccessFile> file_;
+  const std::unique_ptr<RandomAccessFile> file_;
+  const size_t alignment_;
+  const size_t readahead_size_;
+
+  mutable std::mutex lock_;
+  mutable AlignedBuffer buffer_;
+  mutable uint64_t buffer_offset_;
+};
+
+class ReadaheadSequentialFile : public SequentialFile {
```

Comment:
Will take a look @siying

Reply:
@siying I am not sure I understand the proposal. 

I followed the pattern of NewReadaheadRandomAccessFile, which returns RandomAccessFile, which can later be used with RandomAccessFileReader.

Similarly, we now have NewReadaheadSequentialFile, which returns SequentialFile, which can be used with the existing SequentialFileReader (assuming use_direct_io is false, as otherwise it fallback to PositionedRead). All stats, rate limiting, etc. can be added to the existing SequentialFileReader.

So how is your proposal different from what I already have?

Reply:
I would create a different class, rather than a class inheriting SequenceFile. That's it:) Maybe it's as simple as remove the "public SequenceFile" and change the class name in some places when it is used.

Reply:
@elipoz sorry I didn't see this comment when I posted the other comment. I hope we can remove ReadaheadRandomAccessFile. The fact that it is a wrapper and implemented RandomAccessFile caused several problems in the past. One instance I remembered was that we changed the meaning of a function in RandomAccessFile() and forgot that it would break ReadaheadRandomAccessFile as it also implements the function.

Reply:
Hmmm... Stacking functionality without changing interface is a common design pattern in C++...

The SequentialFileReader class already exists. Should I add a new class - ReadaheadSequentialFileReader (that doesn't inherit anything)? Or should I pack all the read-ahead logic into existing SequentialFileReader?

Either approach seems a bit awkward since the former creates plurality of XXXFileReader classes (which do not have common interface, if we do not use virtual methods), while the latter creates a class with too much logic in it that is hard to follow, having to deal with read-ahead, direct-io, and other features (e.g., it already has ROCKSDB_LITE codepath).

Please advise, since I do not see which one is better.

Reply:
"readahead logic to there will be the best solution" -  replied to this already. There are certain complexities to it. See above.
Anyway, the latest version hides the creation of new ReadaheadSequenceFile.